### PR TITLE
Loaders: load JSON and Yaml files

### PR DESCRIPTION
**Pull request by rantonmattei**
Load JSON, Yaml or PhpArray
For files it is using [Symfony's FIleLoader](https://github.com/symfony/Config/blob/master/Loader/FileLoader.php). The `FileLoaderAbstract` extends that class.

Note: `FileLoaderAbstract` uses late static binding to get the extensions from the child class.

This is used to load the logger config from various resources.

Will add .ini support in the future

@mortaliorchard @OrCharles

### CODE REVIEW

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Validate a file extension against a list of provided valid extensions
+     *
+     * @param  string $filepath file path of the file we want to check
+     * @return boolean whether or no the extension is valid
+     */
+    public function validateExtension($filepath)
+    {
+        foreach (static::$validExtensions as $ext) {
+            if ($ext === pathinfo($filepath, PATHINFO_EXTENSION)) {
```
Comment (mortaliorchard) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: Why not just using `in_array`?

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+     *
+     * @return array | mixed   return the section of an array or just a value
+     */
+    public function getSectionOf($array, $section = '')
+    {
+        if (!empty($section) && array_key_exists($section, $array)) {
+            return $array[$section];
+        }
+
+        return $array;
```
Comment (mortaliorchard) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: it seems like getSectionOf will return the array itself if the condition is not matched?

File: src/Config/Loader/FileLoader/Json.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     * exception when we actually load it.
+     *
+     * @param  string  $string String to evaluate
+     * @return boolean whether or not the passed string is meant to be a JSON string
+     */
+    private function isJson($string)
+    {
+        return (
+            !empty($string) &&
+            ($string{0} === '[' || $string{0} === '{')
```
Comment (mortaliorchard) [inline src/Config/Loader/FileLoader/Json.php]: non blocking - just for my own knowledge. what's the difference between `$string[0]` and `$string{0}`

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php (hunk 2 above)
Comment (rantonmattei) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: Yeah, it was a choice on my end. I don't think it is worth throwing an error here. It is either: return the entire array or null.

I need to revisit this at some point to add support for multiple level.

File: src/Config/Loader/FileLoader/Json.php (hunk 3 above)
Comment (rantonmattei) [inline src/Config/Loader/FileLoader/Json.php]: They are similar.
http://php.net/manual/en/language.types.string.php#language.types.string.substr

Although they both work, I think using `[]` might be more confusing as it is generally used for arrays.
I think $str{$index} gives more context.

It is kind of arbitrary I guess, there nothing specified in our standards for this.
My rule of thumb was
`$a[0]` ==> `$a` is an array
`$a{0}`==> `$a` is a string


note that (and this is where Php is confusing), you can use `{}` to evaluate an expression
`$myObj->{$someDynamicProperty}` where `$someDynamicProperty` is set to some string.

Comment (OrCharles) [inline src/Config/Loader/FileLoader/Json.php]: yeah, to Raph's point re variable variables:
http://php.net/manual/en/language.types.string.php#language.types.string.parsing.complex

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php (hunk 1 above)
Comment (rantonmattei) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: Yeah, you're right. Changed it

File: src/Config/Loader/FileLoader/Json.php (hunk 3 above)
Comment (mortaliorchard) [inline src/Config/Loader/FileLoader/Json.php]: Thanks!

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            $input = file_get_contents($input);
+        }
+
+        return trim($input);
+    }
+
+    /**
+     * Test if a given resource is a file name or a file path
+     *
+     * @param  string $resource Plain strig or file path
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: string

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+     * @return boolean whether or not the resource is a file
+     */
+    public function isFile($resource)
+    {
+        return (strpos($resource, "\n") === false) && is_file($resource);
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: interesting check on newline, can you explain why?

**Comment by rantonmattei** [comment]
Anything else, guys?

File: src/Config/Loader/FileLoader/Json.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+     * Valid file extensions for this loader
+     * @var array
+     */
+    public static $validExtensions = array('json');
+
+    /**
+     * Load a JSON string/file
+     *
+     * @param  string $resource JSON string or file path to a JSON file
+     * @param  string|null $type This is not used.
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/Json.php]: curious: why is this here if it's not used?

File: src/Config/Loader/FileLoader/Json.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+     * @return array array containing data from the parse JSON string or file
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/Json.php]: parsed*

File: src/Config/Loader/FileLoader/Json.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    private function isJson($string)
+    {
+        return (
+            !empty($string) &&
+            ($string{0} === '[' || $string{0} === '{')
+        );
+    }
+
+    /**
+     * Return whether or not the passed in resrouce is supported by this loader
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/Json.php]: resource*

File: src/Config/Loader/FileLoader/Json.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+     *
+     * @param  string $resource plain string or filepath
+     * @param  string $type     not used
```
Comment (OrCharles) [inline src/Config/Loader/FileLoader/Json.php]: again I'm curious about the type.. what is it for if it's not used? :dancer:

**Comment by OrCharles** [comment]
looks like a nice clean base: are you intending on doing some (not in a bad way) magic to auto-determine(cascade) what type of config/file you'll be loading?

**Comment by OrCharles** [comment]
just a couple of comment typos and a few questions.. R~

File: src/Config/Loader/FileLoader/FileLoaderAbstract.php (hunk 5 above)
Comment (rantonmattei) [inline src/Config/Loader/FileLoader/FileLoaderAbstract.php]: I'm lazy, I reused what was in the Yaml package in Symfony...
https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Yaml/Yaml.php#L58

Since you can pass either a filepath or a Yaml string, if you see a new line, it is not a file name. so no need to check for `is_file`. It is a cheap way of differenciating those 2

File: src/Config/Loader/FileLoader/Json.php (hunk 6 above)
Comment (rantonmattei) [inline src/Config/Loader/FileLoader/Json.php]: Because it is kind of redundant with the `support` function. Also, in my case, I need to load both strings and JSON. I did not want to have to explicitly specify `"string"` or file everytime. That code would have been repeated in the Yaml loader as well. If I had used it the Loader classes, the code in from the `readFrom` function in `FileLoaderAbstract` class would have been duplicated in those 2 classes as well. I essentially check if this is a file. If it is I read the file (regardless of the extension), if it is not I just return a trimmed string.
I went back and forth a few times on this and I did not want to duplicate code in those classes. I wanted them to be minimal

Comment (rantonmattei) [inline src/Config/Loader/FileLoader/Json.php]: For the type, I think a straight forward use case is that if I had only one load all kind of class and you can have something like a switch statement with the type, etc. But since I already had more granularity, I figured this was not useful for me.

**Comment by rantonmattei** [comment]
Regarding your previous comment, those loaders are eventually be used by what Symfony calls a Resolver. It basicallyu figure out what loader to use to load a passed in resource.

See in what's coming next:
```php
/**
 * Loader class that loads Yaml, JSON and array from various resources (file, php array, string)
 * @see DelegatingLoader
 *
 * @author Raphael Antonmattei <rantonmattei@theorchard.com>
 */
class ConfigLoader extends DelegatingLoader
{
    /**
     * Locator
     * @var Symfony\Component\Config\FileLocator
     */
    protected $locator = null;

    /**
     * Instantiate a Loader object
     * @todo: have the locator passed to the constructor so we can load more than one file
     */
    public function __construct()
    {
        $this->locator = new FileLocator();

        $loaderResolver = new LoaderResolver(array(
            // Do not change that order, it does matter as the resolver returns the first loader
            // that meets the requirements of the "supports" method for each of those loaders
            new ArrayLoader(),
            new JsonLoader($this->locator),
            new YamlLoader($this->locator)
        ));

        parent::__construct($loaderResolver);
    }

    /**
     * Loads a configuration resource: file, array, string
     *
     * @param  mixed $resource resource to load
     * @param  mixed $type not used
     * @return array array of config options
     */
    public function load($resource, $type = null)
    {
        return parent::load($resource);
    }
}
```

**Comment by rantonmattei** [comment]
It looked to me over-engineered at first but it is now really easy to add any file format and we can add more features in the future. And it was fun to learn about those Symfony components

**Comment by rantonmattei** [comment]
Fixed the typos

**Comment by OrCharles** [comment]
R+
